server/src/main/java/org/elasticsearch/script/field/WriteField.java

```
@@ -0,0 +1,403 @@
+        }
+
+        return predicate.test(value);
+    }
+
+    // Value Update
+
+    /**
+     * Update each value at this path with the {@param transformer} {@link Function}.
+     */
+    @SuppressWarnings("unchecked")
+    public WriteField transform(Function<Object, Object> transformer) {
+        if (leaf == null) {
+            return this;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return this;
+        }
+
+        if (value instanceof List<?> list) {
+            ((List<Object>) list).replaceAll(transformer::apply);
+        } else {
+            container.put(leaf, transformer.apply(value));
+        }
+
+        return this;
+    }
+
+    // Value Delete
+
+    /**
+     * Remove all duplicate values from this path.  List order is not preserved.
+     */
+    @SuppressWarnings("unchecked")
+    public WriteField deduplicate() {
+        if (leaf == null) {
+            return this;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return this;
+        }
+
+        if (value instanceof List<?> list) {
+            // Assume modifiable list
+            Set<Object> set = new HashSet<>(list);
+            list.clear();
+            ((List<Object>) list).addAll(set);
+        }
+
+        return this;
+    }
+
+    /**
+     * Remove all values at this path that match {@param filter}.  If there is only one value and it matches {@param filter},
+     * the mapping is removed, however empty Lists are retained.
+     */
+    public WriteField removeValuesIf(Predicate<Object> filter) {
+        if (leaf == null) {
+            return this;
+        }
+
+        Object value = container.getOrDefault(leaf, MISSING);
+        if (value == MISSING) {
+            return this;
+        }
+
+        if (value instanceof List<?> list) {
+            list.removeIf(filter);
```

assert for mutable list here too.

---

My thinking is we should let it fail in the standard way, what do you think?